Marketing Connections: Add Mastodon

fmfernandes opened the pull request:
#### Proposed Changes

* This adds the option to connect to a Mastodon instance.

![image](https://user-images.githubusercontent.com/24195829/216421289-b0ee793c-2102-4718-9446-45667ad80acb.png)


#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* This should be tested alongside D93584-code (which also needs to be reviewed).
* Go to `/marketing/connections/yoursite.com`.
* Try to connect to a Mastodon instance.

Related discussion and more detailed testing instructions: pewaj1-3x-p2

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

fmfernandes committed:
Add service description

fmfernandes committed:
Update config files to support Mastodon feature

fmfernandes committed:
Update placeholder count if Mastodon is enabled

fmfernandes committed:
Add Mastodon to Connections

fmfernandes committed:
Disable button if a connection is being made

mpkelly reviewed (commented):
I have reviewed this and it looks fine. I get that you followed the existing code style, but we try to write new code in TypeScript where possible. So ideally we'd have `mastodon.tsx` with some types in that file. I also think `isValidUsername` is worth unit testing.  

One issue I found is that if you enter an invalid username and then a valid one, the error message is not cleared and it doesn't allow you to connect until you blur the input. Here is a video:

https://user-images.githubusercontent.com/6851384/221825804-e4670afc-446c-4975-b9e0-228bec49615f.mp4


It would also be good to show some kind of feedback that the username validation was in progress i.e. a spinner at the right side of the `input`. 

The popup window opens, and I see the login form. I don't have an account, so I will leave that for others to test. Everything happens quite slowly - it took over a minute to see the login form - and I am not sure how much of that is down to being sandboxed. I applied D93584 but didn't review that code. 

<img width="768" alt="Screenshot 2566-02-28 at 17 25 42" src="https://user-images.githubusercontent.com/6851384/221827017-9e3650c0-53d3-4fde-95ca-aa35f73d74c3.png">

fmfernandes committed:
Change component to use TypeScript

fmfernandes committed:
Add tests for the Mastodon component

fmfernandes committed:
Add tests for `isValidUsername`

fmfernandes committed:
Merge branch 'trunk' into add/mastodon

mpkelly commented:
Hi @fmfernandes. Nice job reworking this PR. Everything is much faster now for me. I also created an account so I can test e2e. Here is some more feedback:

1. You shouldn't need `useEffect` or `useCallback` in `mastodon.tsx`. The top part can be reworked to this:

```tsx
	const translate = useTranslate();
	const [ instance, setInstance ] = useState( '' );
	const [ error, setError ] = useState( '' );

	/**
	 * Checks if username matches a valid Mastodon username.
	 */
	const isValidUsername = ( username: string ) =>
		/@?\b([A-Z0-9_]+)@([A-Z0-9.-]+\.[A-Z]{2,})\b/gi.test( username );

	const handleInstanceChange = ( e: ChangeEvent< HTMLInputElement > ) => {
		const instance = e.target.value.trim();
		setInstance( instance );
		if ( isValidUsername( instance ) || ! instance ) {
			setError( '' );
		} else {
			setError( translate( 'This username is not valid.' ) );
		}
	};
```
I tried the above locally and everything seems to work still. 

2. I don't know the Mastodon username format but I think we are accepting invalid values here:

![image](https://user-images.githubusercontent.com/6851384/222336228-21cc5e2a-5846-48fd-b330-ddab15493a08.png)

3. I can now connect my test account. It tells me that "Jetpack" wants permission to access my account - is this correct, or should it be WordPress.com?

<img width="769" alt="Screenshot 2566-03-02 at 11 51 49" src="https://user-images.githubusercontent.com/6851384/222336512-dfb0a187-668c-414b-805a-9d83683c6b60.png">

4. I hit "authorize" and I get an error: "Missing access token". I have sandboxed the API and applied `D93584`. I am running your Calypso branch locally. Do I need to do anything else to test this?

![image](https://user-images.githubusercontent.com/6851384/222336620-cd9ccb45-6d65-4f57-ae1f-2cad274d4f75.png)

Please tag me in any replies, and I will try and retest and approve ASAP.

fmfernandes committed:
Update instance input handler

It was a little bit overcomplicated and not necessary at all.
It's now simpler and should handle pasting in the input as well.

mpkelly commented:
Hi, @fmfernandes, React hooks can be quite complicated. When they first arrived, I used `useCallback` everywhere, but it's not really needed much in practice. 

The code looks good now. I did some more testing and also created a second account. Here's what I found:

1. I was able to connect my `@a8ctest` account today and there was no error with the access token. 👍 As I had already authorised the connection yesterday, the popup window opened and then closed automatically. I was then shown this message in Calypso:

![image](https://user-images.githubusercontent.com/6851384/222657893-89086167-b1c1-4200-a0b5-5ad0ccedb639.png)

I hit connect and then I see my connection is listed below:

<img width="1079" alt="Screenshot 2566-03-03 at 13 42 50" src="https://user-images.githubusercontent.com/6851384/222658114-85364b03-8485-4981-a0a0-53b2155ee6d4.png">

I then created a new account and tried to add that. 

2. Using my new account, the popup opened and closed automatically. For a new account, I would expect to see the Mastodon "authorize" page but I didn't. On returning to Calypso I saw this toast message only:

![image](https://user-images.githubusercontent.com/6851384/222658485-6d62a8b4-3d34-4655-9e50-28199f173d42.png)

But the second account I added is not listed and doesn't appear to be connected. 

3. My new WP posts don't appear in Mastodon. I checked both of my accounts, and nothing appears. I made one post before the site was launched and another after launching, but I don't see anything [in Mastodon](https://mastodon.social/@a8ctest). 

4. I tested disconnecting/reconnecting and it appears to work ok. 

5. I added a second user to my site as an editor, and I was able to see the connection. However, as the editor, I was able to uncheck the "Connection available to all administrators, editors, and authors" checkbox which feels a bit wrong. Maybe this checkbox should only be shown if you established the connection. 

One other thing, with this checkbox unchecked, the editor can go to **Marketing** > **Connections** and click connect but when I do that I see this error in the popup:

![image](https://user-images.githubusercontent.com/6851384/222661333-036ab659-02d1-4270-bd1a-9e713425ed74.png)

Side note: I think we need a similar checkbox for GH integration. We currently share the connection with other admins but this might be appropriate - so thanks for the idea.

mpkelly commented:
>Woohoo! Thanks for that @mpkelly, I didn't catch this one but I fixed it. Added force_login=False as per the [documentation](https://docs.joinmastodon.org/methods/oauth/#query-parameters) so it's now possible to add multiple accounts from the same instance.

I tried with the two accounts I used previously. I disconnected both of them and then tried to reconnect, but I never the authorisation page in the popup, which automatically closes. For the GH integration, we added an action that listens for the keyring connection being deleted and then deletes the app installation at GitHub. I think something like that would be appropriate here, too. You can see what we did in `Keyring_Service_Github_Deploy#keyring_token_deleted`. 

Even with the above behaviour, I cannot get two connections working for the same user on the same site. The UI suggests this should be supported. I can connect one Mastodon account, but when I connect the second, it doesn't get added. In an earlier test,  I saw a toast message saying it was added, but it didn't appear, then, when I retested and recorded it, I didn't see anything - the popup closes and nothing happens:

https://user-images.githubusercontent.com/6851384/224254970-1401fb0f-7d20-4f8c-bb9e-0a56731b1ee3.mp4

I can get multiple connections working, but I need to use two accounts - my admin and the editor I invited. Both of these can establish a single connection and then check the "Connection available to all administrators, editors, and authors" checkbox to share with each other:

![image](https://user-images.githubusercontent.com/6851384/224255408-330e804d-e825-4e69-b8ef-3b7d028eb739.png)

So the API can return multiple connections, but it looks like it still needs work for a single user to have multiple connections (assuming that is supported). 

>Should we create a separate issue for this?

Ok, I will create one. 

>Could you explain the steps you took to get there? The only way I can see that happening is if you tested that while sandboxing just the API on WordPress.com where the instance input will not be present.

Somehow I had jumped from calypso.local:3000 to wordpress.com during testing. So you can ignore this one.

fmfernandes committed:
Merge branch 'trunk' into add/mastodon

mpkelly commented:
Hi @fmfernandes, I revoked access to Mastodon for both my accounts in the Mastodon account settings and then tried to add two accounts for the same WP user to the same site. Here's a video:

https://user-images.githubusercontent.com/6851384/225205278-babbb68d-bc7e-467c-8042-82df7496fd79.mp4

What I realised was even though the second account is `@kikemelly@mastodon.social`, it still tries to authorise under the first account because I am still signed into Mastodon with that account. So ideally, there would be a way to force a login for the second account. It does work if I log out of the first account and then try to connect `@kikemelly@mastodon.social`. 

On authorising the second account, the Calypso dialog is a bit confusing. It asks me to choose an account, but I can't change anything as there is a single radio button that is already selected:

![image](https://user-images.githubusercontent.com/6851384/225206335-293b591d-ae10-4d38-8b66-c2f1c80b093f.png)

mpkelly commented:
> Do you think that asking for _Enter your Mastodon **username**_ and expecting `@username@example.com` is adding up to the confusion?

Yes, this part is misleading, IMO. It seems like removing the input and just adding a connect button would make the most sense. This way the user is still expecting to enter the account they want to connect. We'd also have less code to maintain and wouldn't need to validate the input. 

I am going to approve this now anyway. The above is just a suggestion.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7880254

Thank you @fmfernandes for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7880254).

manzoorwanijk commented on client/my-sites/marketing/connections/mastodon.tsx:
```diff
@@ -0,0 +1,144 @@
+	const setInstanceToConnectURL = () => {
+		const url = new URL( service.connect_URL );
+		url.searchParams.set( 'instance', instance );
+		service.connect_URL = url.toString();
```
We should not be mutating props here. CC @Automattic/team-calypso

tyxla replied:
Agreed - this is worth a follow-up.
